src/lib/validate-declarative-schema.js

```diff
@@ -104,8 +104,12 @@ const TaskSchema = joi.array().items(
     contactLabel:
       joi.alternatives().try( joi.string().min(1), joi.function() ).optional()
       .error(taskError('"contactLabel" should either be a non-empty string or function(contact, report)')),
-    resolvedIf: joi.function().required()
-      .error(taskError('"resolvedIf" should be of type function(contact, report)')),
+    resolvedIf: joi.alternatives().conditional('actions', {
+      is: joi.array().items(joi.object({ type: 'contact' }).unknown()),
+      then: joi.function().required(),
+      otherwise: joi.function().optional()
+    })
+      .error(taskError('"resolvedIf" should be of type function(contact, report) if there is no report action')),
```

Comment:
There can be multiple actions and it is sufficient to have one action with `type: report`. How do we convey it or we don't need to write everything in the error message?

Reply:
I'm not sure if I understand. You mean that you're getting this error for actions that are not 'report' ?
If that's the case, can you try [.when()](https://joi.dev/api/?v=17.4.0#anywhencondition-options)?

Reply:
I was talking about how we phrase the error message. My original error message was:
> ... if there is no report action

Your suggestion:
> ... if action type is not "report"

Your message is clearer but it does not say that there can be multiple actions and we only need one of them to be of report type.

```js
var resolvingForm = taskDefinition.actions.find(function (action) { return action.type === 'report'; }).form
```

Reply:
Oh I see, well "resolvedIf" should always be a function, so its type will be the same. 
What's different is if "resolvedIf" is mandatory or not when defining each action.

> Schema error in actions array: Actions with property "type" which value is different than "report", should define property "resolvedIf" as: function(contact, report) { ... }.

This is very specific, what do you think?  🤔

Reply:
Yes, it is good, but we might also have to consider that `type` is also optional.

Reply:
#### `resolvedIf` is optional in these cases:
1. One action with type: `report`
```js
actions: [
            {
                type: 'report',
                form: 'pregnancy_followup',
                label: 'Pregnancy followup'
            }
        ],
```

2. Two actions: one is `contact` type, another is `report`
```js
actions: [
            {
                type: 'contact',
            },
            {
                type: 'report',
                form: 'pregnancy_followup',
                label: 'Pregnancy followup'
            }
        ],
```
3. One action, type is not defined, defaults to report
```js
actions: [
            {
                form: 'pregnancy_followup',
                label: 'Pregnancy followup'
            }
        ],
```

#### `resolvedIf` is mandatory in this case:
1. One action with type: 'contact'
```js
actions: [
            {
                type: 'contact',
                label: 'New contact'
            }
        ],
```

Reply:
Nice! This last comment is good and nice to add in the ticket, it'll help QA when testing the feature. Also for any training doc out there. 

Regarding the error message, it still applies since when the `action.type` is missing then it's defaulting to a value.

However, if historically this leads to a frequent struggle, then we can warn the user that an `action.type` is going to default to a value, when processing the tasks (before upload of settings). So it's a different message, for example:

`WARN: Target's action is missing the property type, will default to [something]`

`ERROR: Schema error in actions array: Actions with property "type" which value is different than "report", should define property "resolvedIf" as: function(contact, report) { ... }.`

What do you think? :)

Reply:
Sounds good, where do you suggest putting the warning?

Reply:
I wonder if we can do something in the Joi validation using message() and warning() or perhaps using a combination with when().then() 🤔 Otherwise maybe just after the Joi schema validation, we can create a function that checks and print the warning.